### PR TITLE
Feature : Form(Entity, Repository.Contract, Repository, RepositoryTests)

### PR DESCRIPTION
**Pull request by Ellyria34**
CRUD methods, documentation and unit tests done.

### CODE REVIEW

**Review by htkisui — APPROVED** [review]

File: QuikForm.Repositories/Repositories/FormRepository.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    private readonly ApplicationDbContext _context;
+
+    public FormRepository(ApplicationDbContext context)
+    {
+        _context = context;
+    }
+
+    public async Task CreateAsync(Form form)
+    {
+        await _context.AddRangeAsync(form);
```
Comment (htkisui) [inline QuikForm.Repositories/Repositories/FormRepository.cs]: Use AddAsync instead.

File: QuikForm.Repositories/Repositories/FormRepository.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        await _context.SaveChangesAsync();
+    }
+
+    public async Task DeleteAsync(int id)
+    {
+        Form formToDelete = await _context.Forms.FirstOrDefaultAsync(i => i.Id == id)?? throw new FormNotFoundException();
```
Comment (htkisui) [inline QuikForm.Repositories/Repositories/FormRepository.cs]: "i" is ambiguous, use "f" for Form instead.

File: QuikForm.Repositories/Repositories/FormRepository.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    public async Task<List<Form>> GetAllAsync()
+    {
+        return await _context.Forms.ToListAsync();
+    }
+
+    public async Task<Form> GetByIdAsync(int id)
+    {
+        Form form = await _context.Forms.FirstOrDefaultAsync(i => i.Id == id) ?? throw new FormNotFoundException();
```
Comment (htkisui) [inline QuikForm.Repositories/Repositories/FormRepository.cs]: "i" -> "f"

File: QuikForm.Repositories/Repositories/FormRepository.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        return form;
+    }
+
+    public async Task<Form> UpdateAsync(Form form)
+    {
+        Form formToUpdate = await _context.Forms.FirstOrDefaultAsync(i=>i.Id == form.Id)?? throw new FormNotFoundException();
```
Comment (htkisui) [inline QuikForm.Repositories/Repositories/FormRepository.cs]: "i" -> "f"

File: QuikForm.Repository.Contracts/Contracts/IFormRepository.cs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    /// </summary>
+    /// <param name="id"></param>
+    /// <returns>The form.</returns>
+    Task<Form> GetByIdAsync(int id);
+
+    /// <summary>
+    /// Update a form by id.
+    /// </summary>
+    /// <param name="form"></param>
+    /// <returns></returns>
```
Comment (htkisui) [inline QuikForm.Repository.Contracts/Contracts/IFormRepository.cs]: Returns value missing

File: QuikForm.Repository.Contracts/Exceptions/Forms/FormNotFoundException.cs [hunk 6]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace QuikForm.Repository.Contracts.Exceptions.Forms;
+public class FormNotFoundException : Exception
+{
+    public override string Message => "Form is not Found.";
```
Comment (htkisui) [inline QuikForm.Repository.Contracts/Exceptions/Forms/FormNotFoundException.cs]: Found => found

File: QuikForm.Repository.Test.Units/Repositories/FormRepositoryTests.cs [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+
+namespace QuikForm.Repositories.Repositories.Tests;
+
+[TestClass()]
+public class FormRepositoryTests
+{
+    #region CreateAsync
+
+    [TestMethod()]
+    public async Task CreateAsync_FormToAdd_FormWithId()
```
Comment (htkisui) [inline QuikForm.Repository.Test.Units/Repositories/FormRepositoryTests.cs]: CreateAsync_FormToAdd_FormAdded

File: QuikForm.Repository.Test.Units/Repositories/FormRepositoryTests.cs [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+    {
+        // Arrange
+        DbContextOptionsBuilder<ApplicationDbContext> builder = new DbContextOptionsBuilder<ApplicationDbContext>().UseInMemoryDatabase("QuikFormTest");
+        ApplicationDbContext context = new ApplicationDbContext(builder.Options);
+        context.Database.EnsureDeleted();
+        FormRepository formRepository = new FormRepository(context);
+        int IdExpected = 1;
```
Comment (htkisui) [inline QuikForm.Repository.Test.Units/Repositories/FormRepositoryTests.cs]: idExpected

File: QuikForm.Repository.Test.Units/Repositories/FormRepositoryTests.cs [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+        await context.Forms.AddAsync(new Form() { Id = 1, Title = "Title1", Description = "Description", PublishedAt = DateTime.Parse("12/01/2024"), ClosedAt = DateTime.Parse("24/01/2024") });
+        await context.SaveChangesAsync();
+        Form formExpected = new Form { Id = 1, Title = "TitleUpdate", Description = "DescriptionUpdate", PublishedAt = DateTime.Parse("15/01/2024"), ClosedAt = DateTime.Parse("25/01/2024") };
+
+        // Act
+        Form formResult = await formRepository.UpdateAsync(formExpected);
+        ApplicationDbContext otherContext = new ApplicationDbContext(builder.Options);
+
+        // Assert
+        Assert.AreEqual(formExpected.Title, otherContext.Forms.Find(1)?.Title);
```
Comment (htkisui) [inline QuikForm.Repository.Test.Units/Repositories/FormRepositoryTests.cs:162]: description, publishedAt, closedAt Asserts missing.

File: QuikForm.Repository.Test.Units/Repositories/FormRepositoryTests.cs [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+    }
+
+    [TestMethod()]
+    [ExpectedException(typeof(FormNotFoundException))]
+    public async Task UodateAsync_InvalidFormToSearch_ThrowFormNotFoundException()
```
Comment (htkisui) [inline QuikForm.Repository.Test.Units/Repositories/FormRepositoryTests.cs]: UpdatedAsync_InvalidForm_ThrowFormNotFoundException

File: QuikForm.Repository.Test.Units/Repositories/FormRepositoryTests.cs [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+    public async Task UpdateAsync_FormToUpdate_FormUpdated()
+    {
+        // Arrange
+        DbContextOptionsBuilder<ApplicationDbContext> builder = new DbContextOptionsBuilder<ApplicationDbContext>().UseInMemoryDatabase("QuikFormTest");
+        ApplicationDbContext context = new ApplicationDbContext(builder.Options);
+        context.Database.EnsureDeleted();
+        FormRepository formRepository = new FormRepository(context);
+        await context.Forms.AddAsync(new Form() { Id = 1, Title = "Title1", Description = "Description", PublishedAt = DateTime.Parse("12/01/2024"), ClosedAt = DateTime.Parse("24/01/2024") });
+        await context.SaveChangesAsync();
+        Form formExpected = new Form { Id = 1, Title = "TitleUpdate", Description = "DescriptionUpdate", PublishedAt = DateTime.Parse("15/01/2024"), ClosedAt = DateTime.Parse("25/01/2024") };
```
Comment (htkisui) [inline QuikForm.Repository.Test.Units/Repositories/FormRepositoryTests.cs]: new Form()

File: QuikForm.Repository.Test.Units/Repositories/FormRepositoryTests.cs [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+    public async Task UodateAsync_InvalidFormToSearch_ThrowFormNotFoundException()
+    {
+        // Arrange
+        DbContextOptionsBuilder<ApplicationDbContext> builder = new DbContextOptionsBuilder<ApplicationDbContext>().UseInMemoryDatabase("QuikFormTest");
+        ApplicationDbContext context = new ApplicationDbContext(builder.Options);
+        context.Database.EnsureDeleted();
+        FormRepository formRepository = new FormRepository(context);
+
+        // Act
+        Form form = new Form();
```
Comment (htkisui) [inline QuikForm.Repository.Test.Units/Repositories/FormRepositoryTests.cs]: Add a least the Id

**Review by htkisui — COMMENTED** [review]
naming issues

File: QuikForm.Repositories/Repositories/FormRepository.cs (hunk 1 above)
Comment (Ellyria34) [reply QuikForm.Repositories/Repositories/FormRepository.cs]: done

File: QuikForm.Repositories/Repositories/FormRepository.cs (hunk 2 above)
Comment (Ellyria34) [reply QuikForm.Repositories/Repositories/FormRepository.cs]: done

File: QuikForm.Repositories/Repositories/FormRepository.cs (hunk 3 above)
Comment (Ellyria34) [reply QuikForm.Repositories/Repositories/FormRepository.cs]: done

File: QuikForm.Repositories/Repositories/FormRepository.cs (hunk 4 above)
Comment (Ellyria34) [reply QuikForm.Repositories/Repositories/FormRepository.cs]: done

File: QuikForm.Repository.Contracts/Contracts/IFormRepository.cs (hunk 5 above)
Comment (Ellyria34) [reply QuikForm.Repository.Contracts/Contracts/IFormRepository.cs]: done

File: QuikForm.Repository.Test.Units/Repositories/FormRepositoryTests.cs (hunk 8 above)
Comment (Ellyria34) [reply QuikForm.Repository.Test.Units/Repositories/FormRepositoryTests.cs]: done

File: QuikForm.Repository.Test.Units/Repositories/FormRepositoryTests.cs (hunk 9 above)
Comment (Ellyria34) [reply QuikForm.Repository.Test.Units/Repositories/FormRepositoryTests.cs:162]: done

File: QuikForm.Repository.Test.Units/Repositories/FormRepositoryTests.cs (hunk 10 above)
Comment (Ellyria34) [reply QuikForm.Repository.Test.Units/Repositories/FormRepositoryTests.cs]: done

File: QuikForm.Repository.Test.Units/Repositories/FormRepositoryTests.cs (hunk 11 above)
Comment (Ellyria34) [reply QuikForm.Repository.Test.Units/Repositories/FormRepositoryTests.cs]: done

File: QuikForm.Repository.Test.Units/Repositories/FormRepositoryTests.cs (hunk 12 above)
Comment (Ellyria34) [reply QuikForm.Repository.Test.Units/Repositories/FormRepositoryTests.cs]: done

**Comment by Ellyria34** [comment]
Done

**Review by htkisui — APPROVED** [review]

File: QuikForm.Repository.Contracts/Exceptions/Forms/FormNotFoundException.cs (hunk 6 above)
Comment (Ellyria34) [reply QuikForm.Repository.Contracts/Exceptions/Forms/FormNotFoundException.cs]: done

File: QuikForm.Repository.Test.Units/Repositories/FormRepositoryTests.cs (hunk 7 above)
Comment (Ellyria34) [reply QuikForm.Repository.Test.Units/Repositories/FormRepositoryTests.cs]: done

**Pull request closed without merging**